### DISCUSSION TITLE
Why do i need Symbiote? There are tons of JavaScript libraries, are they not enough?

### DISCUSSION THREAD

**Discussion by foxeyes** [symbiotejs/symbiote.js#13] [Q&A]
Symbiote is designed around of set of specific requirements of the micro-frontends architecture and widget building. It created to be organic part of any environment or development stack. 

We spent a lot of time trying to find existing solution for that, but that was failed. The closest solution was [LitElement](https://lit-element.polymer-project.org/guide) but it has a lot of disadvantages for us, such as:
- binding to [Shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) technology (slots and styling are not working without Shadow DOM usage)
- not good for strict [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP) settings (it uses an experimental API for styling, which is working in Chrome browser only, other browsers should have `--unsafe-inline` flag)
- it doesn't have a built in solution for components interoperation and common data context 
- etc...

One of the other considered variants was [Svelte](https://svelte.dev/). It promises the one of the lightest results, which can be wrapped with a standard [Custom Element](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements). But it has a not acceptable dependency to compiler. Also, it has a set of disadvantages from the debugging and testing side. And it doesn't provide any architecture solution for components external communications.

All other mainstream solutions, such as React (~40kb) or Vue.js (~25kb) are just too heavy. We cannot bring such unoptimised things to our host applications in case of light widgets.

**Comment by dragomano** [comment]
Выглядит интересно, добавил в свою [сравнительную табличку](https://php.dragomano.ru/znakomstvo-s-populyarnymi-js-freymvorkami-zaklyuchenie/). Если в чем-то ошибся и упустил какие-то нюансы, буду рад поправкам.

Странно, что русскоязычной документации нет.

**Comment by Shivang9983** [comment]
Here is the summary in exactly three short sentences:The Problem: Mainstream frameworks like React and Vue are too heavy, while alternatives like LitElement and Svelte require complex compilers or cause browser security compatibility issues.The Solution: Symbiote was created as a lightweight, compiler-free alternative tailored specifically for building micro-frontends and widgets.The Benefit: It integrates organically into any development stack without slowing down applications or forcing developers to compromise on component communication.

**Comment by benogoulart** [comment]
I think the key point is that Symbiote isn't trying to replace React, Vue, or Svelte as general-purpose UI frameworks. It's targeting a much more specific use case: lightweight, embeddable widgets and micro-frontends that need to integrate into existing applications with minimal overhead.

For projects where bundle size, CSP compatibility, and framework independence are more important than a rich ecosystem, the trade-offs described here make sense. For larger single-page applications, however, mainstream frameworks still offer significant advantages in terms of tooling, community support, and available libraries.

So the choice really depends on the problem you're trying to solve rather than one framework being universally better than another.
